contracts/src/libraries/verifier/ZkTrieVerifier.sol

```diff
@@ -130,129 +130,121 @@ library ZkTrieVerifier {
-                ptr := add(ptr, 0x01) // skip NodeType
-                require(eq(calldataload(ptr), key), "Node key mismatch")
-                ptr := add(ptr, 0x20) // skip NodeKey
-                {
-                    let valuePreimageLength := and(shr(224, calldataload(ptr)), 0xffff)
-                    // @todo check CompressedFlag
+                switch byte(0, calldataload(ptr))
+                case 4 {
+                    // nonempty leaf node
+                    ptr := add(ptr, 0x01) // skip NodeType
+                    require(eq(calldataload(ptr), key), "AccountKeyMismatch")
+                    ptr := add(ptr, 0x20) // skip NodeKey
+                    require(eq(shr(224, calldataload(ptr)), 0x05080000), "InvalidAccountCompressedFlag")
                     ptr := add(ptr, 0x04) // skip CompressedFlag
-                    ptr := add(ptr, valuePreimageLength) // skip ValuePreimage
-                }
 
-                // compute value hash for State Account Leaf Node
-                {
-                    let tmpHash1 := calldataload(ptr)
-                    ptr := add(ptr, 0x20) // skip nonce/codesize/0
-                    tmpHash1 := poseidon_hash(poseidon, tmpHash1, calldataload(ptr))
+                    // compute value hash for State Account Leaf Node
+                    // [nonce||codesize||0, balance, storage_root, keccak codehash, poseidon codehash]
+                    mstore(0x00, calldataload(ptr))
+                    ptr := add(ptr, 0x20) // skip nonce||codesize||0
+                    mstore(0x00, poseidon_hash(hasher, mload(0x00), calldataload(ptr), 1280))
                     ptr := add(ptr, 0x20) // skip balance
-                    storageHash := calldataload(ptr)
+                    storageRootHash := calldataload(ptr)
                     ptr := add(ptr, 0x20) // skip StorageRoot
-                    let tmpHash2 := hash_uint256(poseidon, calldataload(ptr))
+                    let tmpHash := hash_uint256(hasher, calldataload(ptr), 512)
                     ptr := add(ptr, 0x20) // skip KeccakCodeHash
-                    tmpHash2 := poseidon_hash(poseidon, storageHash, tmpHash2)
-                    tmpHash2 := poseidon_hash(poseidon, tmpHash1, tmpHash2)
-                    tmpHash2 := poseidon_hash(poseidon, tmpHash2, calldataload(ptr))
+                    tmpHash := poseidon_hash(hasher, storageRootHash, tmpHash, 1280)
+                    tmpHash := poseidon_hash(hasher, mload(0x00), tmpHash, 1280)
+                    tmpHash := poseidon_hash(hasher, tmpHash, calldataload(ptr), 1280)
                     ptr := add(ptr, 0x20) // skip PoseidonCodeHash
 
-                    tmpHash1 := poseidon_hash(poseidon, 1, key)
-                    tmpHash1 := poseidon_hash(poseidon, tmpHash1, tmpHash2)
+                    tmpHash := poseidon_hash(hasher, key, tmpHash, 4)
+                    require(eq(leafHash, tmpHash), "InvalidAccountLeafNodeHash")
```

Comment:
The process for hash calculation here is ok. But i advice left more comment or at least some directions (like https://github.com/scroll-tech/mpt-circuit/blob/v0.7/spec/mpt-proof.md) here to make anyone without too much backgroud knowledge are also able to have fast understanding